### PR TITLE
Now use epoll so we can wake up readEvents

### LINKED ISSUE #5: inotify blocks
readEvent() currently blocks.

See https://code.google.com/p/go/issues/detail?id=5536

@howeyc attempted to change this in 3bfa915bde2415222d8dc048945f701353508160 but reverted it 7beb45183d2ab48bfc9616d9a019513a123a4983 due to reports of no new events https://github.com/howeyc/fsnotify/issues/52.

>  This was something I added to stop readEvent from blocking forever.
> Basically calling close would mean no new data came to the file descriptor
> and Read would block forever.
> 
> This was an attempt to use select with a timeout so it would return after
> the timeout stating whether there was new data.

### LINKED ISSUE #40: Cannot remove watch for deleted file
Once a file is deleted, you cannot clean up watches for it. For instance, with fsnotify:

``` go
func (w *Watcher) Remove(name string) error {
	name = filepath.Clean(name)
	w.mu.Lock()
	defer w.mu.Unlock()
	watch, ok := w.watches[name]
	if !ok {
		return fmt.Errorf("can't remove non-existent inotify watch for: %s", name)
	}
	success, errno := syscall.InotifyRmWatch(w.fd, watch.wd)
	if success == -1 {
		return os.NewSyscallError("inotify_rm_watch", errno)
	}
	delete(w.watches, name)
	return nil
}
```

The file is gone, so even though we find the watch in `w.watches` we return a syscall error and don't delete the entry from `w.watches`.

It seems that maybe we should delete from `w.watches` regardless, or at least if `errno` is `EBADF`. If this sounds good, I can make a PR (and try to figure out the appropriate change for other systems (at least darwin) as well.

### PR DESCRIPTION
**Pull request by PieterD**
fixes #5.
fixes #40.
updates #59 (fixed for inotify).

### CODE REVIEW

**Comment by nathany** [comment]
I like how you separated the poller out into a separate file. But man, this makes it so much more complex.

File: inotify_poller.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+// Create a new inotify poller.
+// This creates an inotify handler, and an epoll handler.
+func newFdPoller() (*fdPoller, error) {
+	var errno error
+	poller := new(fdPoller)
+
+	// Create inotify fd
+	poller.fd, errno = syscall.InotifyInit()
```
Comment (nathany) [inline inotify_poller.go]: I think InotifyInit should remain in NewWatcher and pass the fd into newFdPoller.

File: inotify_poller.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+	if poller.fd == -1 {
+		return nil, os.NewSyscallError("inotify_init", errno)
```
Comment (nathany) [inline inotify_poller.go]: I found that wrapping with NewSyscallError breaks comparisons with the underlying error. Ended up removing them in 719fef588d53f4ccdf94803732cc34a6097ce1b7 but I missed this one.

File: inotify_poller.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	}
+	// Create epoll fd
+	poller.epfd, errno = syscall.EpollCreate(1)
+	if poller.epfd == -1 {
+		syscall.Close(poller.fd)
+		return nil, os.NewSyscallError("epoll_create", errno)
+	}
+	// Create pipe; pipe[0] is the read end, pipe[1] the write end.
+	errno = syscall.Pipe(poller.pipe[:])
```
Comment (nathany) [inline inotify_poller.go]: Should this be `err` rather than `errno` in this case?

File: inotify_poller.go (hunk 1 above)
Comment (nathany) [inline inotify_poller.go]: Likewise, NewWatcher should be responsible for closing the inotify fd if poller has an error.

Poller could in theory work for any fd passed in and only clean up its own resources.

File: inotify_poller.go (hunk 3 above)
Comment (nathany) [inline inotify_poller.go]: Maybe for EpollCtl too?

File: inotify_poller.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+	}
+
+	// Register pipe fd with epoll
+	event = syscall.EpollEvent{
+		Fd:     int32(poller.pipe[0]),
+		Events: syscall.EPOLLIN,
+	}
+	errno = syscall.EpollCtl(poller.epfd, syscall.EPOLL_CTL_ADD, poller.pipe[0], &event)
+	if errno != nil {
+		syscall.Close(poller.fd)
```
Comment (nathany) [inline inotify_poller.go]: maybe cleanup could be put in a `defer func() {}` that checks if err is != nil? Might be cleaner, maybe not.

Maybe extracting a few functions, like to create and close the pipe, would help a bit?

File: inotify.go [hunk 5]
```diff
@@ -188,10 +184,14 @@ func (w *Watcher) readEvents() {
 			return
 		}
 
-		// If EOF is received
+		// If EOF is received. This should really never happen.
```
Comment (nathany) [inline inotify.go]: :heart: :hear_no_evil:

File: inotify_poller.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+	}
+
+	return poller, nil
+}
+
+// Wait using epoll, then read from inotify.
+// Returns true if something is ready to be read,
+// false if there is not.
+func (poller *fdPoller) wait() (bool, error) {
+	events := make([]syscall.EpollEvent, 7)
```
Comment (nathany) [inline inotify_poller.go]: why 7?

File: inotify_poller.go [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+		if epollerr {
+			return true, nil
+		}
+		if epollhup || epollpipehup || epollpipein {
+			return false, nil
+		}
+		if epollin {
+			return true, nil
+		}
+		return false, errors.New("Epoll failed to generate any of the only six possibilities.")
```
Comment (nathany) [inline inotify_poller.go]: oh. hm.

**Comment by nathany** [comment]
Should there be inotify_poller_test.go?

File: inotify_poller.go (hunk 6 above)
Comment (PieterD) [inline inotify_poller.go]: There should be 3 possible events for each fd; hangup, error, and input. With two events, that's 6 maximum events if they all come in at once.

I don't know if epoll returns the number of events it wrote to the buffer, or the number of events TOTAL. But I'm catching both by making the buffer one larger than the maximum amount of events.

This should probably be a comment.

File: inotify.go [hunk 8]
```diff
@@ -19,28 +20,39 @@ import (
...
-	done    chan struct{}     // Channel for sending a "quit message" to the reader goroutine
+	Events   chan Event
+	Errors   chan error
+	mu       sync.Mutex // Map access
+	fd       int
+	poller   *fdPoller
+	watches  map[string]*watch // Map of inotify watches (key: path)
+	paths    map[int]string    // Map of watched paths (key: watch descriptor)
+	done     chan struct{}     // Channel for sending a "quit message" to the reader goroutine
+	doneresp chan struct{}     // Channel to respond to Close
```
Comment (nathany) [inline inotify.go]: Maybe `doneResp` instead? Or `doneResposne`?

I was thinking of sending a signal one way over the done channel and then sending one back, but that won't work if the first signal is to close. Looking at the code more closely, I think we need to keep the two separate channels.

File: inotify.go [hunk 9]
```diff
@@ -175,6 +167,20 @@ func (w *Watcher) readEvents() {
 			return
 		}
 
+		ok, errno = w.poller.wait()
+		if errno != nil {
+			select {
```
Comment (nathany) [inline inotify.go]: I'm starting to wonder if we should have a helper function to send an `error` down the channel or err/abort if done.

File: inotify_poller.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+	}
+	errno = syscall.EpollCtl(poller.epfd, syscall.EPOLL_CTL_ADD, poller.pipe[0], &event)
+	if errno != nil {
+		return nil, errno
+	}
+
+	return poller, nil
+}
+
+// Wait using epoll, then read from inotify.
```
Comment (nathany) [inline inotify_poller.go]: ~~then read form inotify~~

File: inotify_poller.go [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+	n, errno := syscall.Read(poller.pipe[0], buf)
+	if n == -1 {
+		return errno
+	}
+	return nil
+}
+
+// Close all poller file descriptors, but not the one passed to it.
+func (poller *fdPoller) close() {
+	syscall.Close(poller.pipe[1])
```
Comment (nathany) [inline inotify_poller.go]: if EpollCreate fails then pipe hasn't been used it. Is this going to break?

File: inotify_poller.go [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+	buf := make([]byte, 1)
+	n, errno := syscall.Write(poller.pipe[1], buf)
+	if n == -1 {
+		return errno
+	}
+	return nil
+}
+
+func (poller *fdPoller) clearWake() error {
+	buf := make([]byte, 100)
```
Comment (nathany) [inline inotify_poller.go]: why 100? :smirk:

File: inotify_poller_test.go [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+	if errno != nil {
+		t.Fatalf("Failed to write to pipe: %v", errno)
+	}
+}
+
+func (tfd testFd) get(t *testing.T) {
+	buf := make([]byte, 10)
+	_, errno := syscall.Read(tfd[0], buf)
+	if errno != nil {
+		t.Fatalf("Failed to write from pipe: %v", errno)
```
Comment (nathany) [inline inotify_poller_test.go]: read

File: inotify_poller.go [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+
+type fdPoller struct {
+	fd   int    // File descriptor (as returned by the inotify_init() syscall)
+	epfd int    // Epoll file descriptor
+	pipe [2]int // Pipe for waking up
+}
+
+// Create a new inotify poller.
+// This creates an inotify handler, and an epoll handler.
+func newFdPoller(fd int) (_ *fdPoller, errno error) {
```
Comment (nathany) [inline inotify_poller.go]: no need for the `_`

Comment (nathany) [inline inotify_poller.go]: Oh. Mixing named and unamed. Mind you errno isn't really needed here for readability of the API. Could move it to var.

Comment (PieterD) [inline inotify_poller.go]: But it is required to reference the return value in the deferred function two lines below.

**Pull request closed without merging**